

\section{Introduction}


Experimental evolution studies have been instrumental in informing our understanding
of the processes shaping evolution \citep{Elena:2003fr}. 
Most of such studies have been carried out in microbes
\citep[reviewed in][]{Dettman:2012gt,Jerison:2015kw}, and have provided insights on
such diverse and fundamental themes as historical contingency, evolutionary innovation,
parallel evolution, and adaptation
\citep{Blount:2008fo,Barrick:2009in,TollRiera:2016fd,Voordeckers:2015cl,Gerstein:2012ki}.


Similar experiments have been carried out in invertebrates
\citep{Gompert:2016eg,Chandler:2014bn,Burke:2010eq,Kang:2016ea,Rouchet:2014jl}, 
although such studies are comparatively rare.
Experimental evolution studies in insects typically utilize limited numbers of 
clonal or inbred lines \citep[e.g.,][]{Rouchet:2014jl,Kang:2016ea} and 
characterize experimental populations by either 
(a) measuring the distribution of specific phenotypes \citep[e.g.,][]{Rouchet:2014jl}
or
(b) sequencing pooled DNA \citep[``Pool-seq''; e.g.,][]{Burke:2010eq}.
The former requires that the researcher manipulate the environment such that a 
specific phenotype is predicted to change.
Moreover, if starting experimental populations have a continuous distribution of
phenotypes or if any significant degree of phenotypic plasticity exists, this method
is not likely to provide accurate estimates of how distributions of individuals change
through time.

Pool-seq, however, is an accurate, cost-effective method to measure allele frequencies
in populations \citep{Gautier:2013dv,Futschik:2010be}
and to identify loci associated with traits \citep{Rubin:2010es,Bastide:2013jx}.
However, Pool-seq's advantageous accuracy-to-cost ratio is only present when there
are many pooled individuals ($>40$) and when depth of coverage is high ($> \times 50$).
Because sequencing error is difficult to distinguish from rare alleles, 
Pool-seq is not ideal when trying to detect these low-frequency alleles
\citep{Schlotterer:2014dk}.
Additionally, Pool-seq of whole-genome sequencing provides much unnecessary 
information if an association study is not the ultimate goal.

One way to reduce genome complexity is to use restriction site-associated DNA 
sequencing (``RADseq'').
RADseq approaches use restriction enzymes to break apart the genome at specific 
locations determined by the enzyme's binding site sequence.
Although some use RADseq to refer to one specific methodology, here I use the more
inclusive definition of RADseq by
\citet{Andrews:2016bc}, who define RADseq as all methods 
``... that rely on restriction enzymes to determine the set of loci to be sequenced''
\citep[p 81]{Andrews:2016bc}.
Genotyping-by-sequencing (``GBS'') is a particularly low-effort, low-cost RADseq approach
that requires no specialized equipment for sample preparation \citep{Elshire:2011gn}.

Herein I will assess pooled GBS as a method to assess the abundance of clonal lines of
aphids after experimental evolution.
Aspects of this specific experiment that make it suitable for pooled GBS are as follows:
\begin{itemize}
    \item Each experimental population will contain $\gg$1,000 individuals when allele 
        frequencies are sought at the end of the experiment.
        Individual-based estimates (e.g., using microsatellites) would require ~50--100 
        individuals to be sampled from the population (See Fig \ref{fig:sampling}), which
        would take huge amounts of preparation time.
    \item 
\end{itemize}



\begin{figure}[!htbp]
    \centering
    \includegraphics[width=6in]{sampling_p.pdf}
    \caption{Simulated estimates of mean genotype abundance in a population for a given
        sample size. Samples were randomly drawn from a population of 1,000 with 
        $\mu = 0.5$. $N$ is the number of samples drawn from the population, and 
        distributions are for 1,000 simulations.
        See \href{https://github.com/lucasnell/mol_ecol/samp_plot.R}{here} for this 
        figure's code.}
    \label{fig:sampling}
\end{figure}



